util/buildinfo/buildinfo.go

```diff
@@ -197,6 +200,46 @@ func decodeDeps(key string, attrs map[string]*string) (map[string]binfotypes.Bui
 	return res, nil
 }
 
+// dedupSources deduplicates regular sources from dependencies ones.
+func dedupSources(bi binfotypes.BuildInfo, depsSources []binfotypes.Source) []binfotypes.Source {
+	if len(bi.Deps) == 0 || len(depsSources) == 0 {
+		return bi.Sources
+	}
+	for i, src := range bi.Sources {
+		for _, dsrc := range depsSources {
+			if src.Type == binfotypes.SourceTypeDockerImage && src.Pin == dsrc.Pin {
```

Comment:
I think we should check direct `src == dsrc` first. For the special check for `SourceTypeDockerImage` ideally we also check the the `src` is a digest only reference.

Eg. lets say in a rare case `alpine:foo` and `alpine:bar` resolve to the same digest. They should not be deduplicated. But if `alpine@sha` and `alpine:bar` return same digest then they should.

Reply:
Ok sgtm, will make that change.

Reply:
Done and added a test for this case https://github.com/moby/buildkit/pull/2684/files#diff-6a10c8a5cc86861c4792a2228b0df8a62e9145c52b939dd0306bc11e04aff6c1R165-R229